b9lab-content/3-main-concepts/10-multistore-keepers.md

```diff
@@ -108,27 +108,357 @@ The AnteHandler is theoretically optional, but still a very important component
+
+In the context of the Cosmos SDK, you need to keep track of where the FIFO starts and stops by saving the corresponding game ids:
+
+```go
+const (
+    NoFifoIdKey = "-1"
+)
+type NextGame struct {
+    ...
+    FifoHead string
+    FifoTail string
+}
+```
+And to have each game know its relative position, and also the number of moves done, to assist the refunding logic on games with 0, 1 or 2+ moves:
+
+```go
+type StoredGame struct {
+    ...
+    MoveCount uint64
+    BeforeId  string
+    AfterId   string
+}
+```
+Next, and this is left as an exercise, you need to code a regular FIFO whereby:
 
-## Long-running exercise
+* Games are sent to the back when created or played on.
+* Games are removed from the FIFO when they are finished or time out.
 
-So far we have described what we wanted to do:
+## When To Expire Games
 
-* Define a game id. How is defined? Probably incrementally.
-* Save many `FullGame`s in storage for retrieval by id.
-* Save `FullGame` ids in an ordered list of their timeouts, for use in `EndBlock`.
-* Handle `CreateGameTx`, `MoveTx`, and `RejectTx`.
+Since you want to expire the games that have timed out at the end of a block, you need to hook your keeper to the right call. When building the whole application, the Cosmos SDK will call at various points into each module. The function it calls at each block's end looks like this:
 
-We need to add the other functions:
+```go
+func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
+    // TODO
+    return []abci.ValidatorUpdate{}
+}
+```
+This is where you write the necessary code, preferably in the keeper. For instance with:
+
+```go
+am.keeper.ForfeitExpiredGames(sdk.WrapSDKContext(ctx))
+```
+Those among you with a well-placed paranoia must be asking whether you can ensure that the execution of this `EndBlock` does not become prohibitively expensive. After all, the number of games to potentially expire is unbounded, which is a recipe for disaster in the blockchain world. Is there a situation or an attack vector that would make this is a possibility?
+
+Fortunately, the timeout duration is fixed and the same for all games. This means that those `n` games that all expire in a given block have all been created or updated at roughly the same time. Or in effect, at roughly the same block height `h` give or take a margin of error `h-1` and `h+1`. These created and updated games are in limited number, as per the validators rules. So if by any chance, all games in blocks `h-1`, `h` and `h+1` expire now, then the `EndBlock` function would have to expire 3 times as many games as a block can handle. This is the worst case scenario, and still sounds manageable.
+
+As a corollary, you should be careful about letting the game creator pick a timeout duration. That would open an avenue for a malicious actor to, for instance, stagger game creations over a large number of blocks, with decreasing timeouts, so that they all expire at the same time.
```

Comment:
Keeping it simple :)
```suggestion
Because of this, you should be careful about letting the game creator pick a timeout duration. That would open an avenue for a malicious actor to, for instance, stagger game creations over a large number of blocks, with decreasing timeouts, so that they all expire at the same time.
```